lib/pbench/server/filetree.py

```diff
@@ -298,20 +327,84 @@ def unpack(self, incoming: Path, results: Path):
         ensured by calling this indirectly through the Controller class unpack
         method.
 
-        TODO: This is a prototype for testing, and doesn't actually unpack
-        the tarball as we're going to be relying on the current unpack pipeline
-        command for some time.
-
         Args:
             incoming: Controller's directory in the INCOMING tree
             results: Controller's directory in the RESULTS tree
+
+        Raises:
+            TarballUnpackError if an exception/failure occurs while unpacking
+            the tarball.
         """
-        self.logger.warning("Tarball unpack is not yet implemented")
-        unpacked = incoming / self.name
-        unpacked.mkdir(parents=True)  # Just create an empty directory for now
-        self.unpacked = unpacked
+        unpacked = incoming / f"{self.name}.unpack"
+        unpacked.mkdir(parents=True)
+
+        try:
+            process = subprocess.run(
+                [
+                    "tar",
+                    "-x",
+                    "--no-same-owner",
+                    "--delay-directory-restore",
+                    "--force-local",
+                    "--file=" + str(self.tarball_path),
+                ],
+                cwd=unpacked,
+                stdin=subprocess.DEVNULL,
+                capture_output=True,
+                text=True,
+            )
+        except Exception as exc:
+            raise TarballUnpackError(self.tarball_path, str(exc)) from exc
```

Comment:
Even if we didn't succeed in unpacking anything at all here, we have still created the `unpacked` temporary directory and we ought to try to remove it before terminating with the exception.

(While I hesitate to even suggest it at this point, this repeated hierarchical cleanup is almost starting to look like something that could use the `Cleanup` action stack manager I built for `PUT`...)